### PR TITLE
Crud on User Profile

### PR DESCRIPTION
**Pull request by midepeter**
Creating lambda functions that trigger upon request through a firebase.This lambda functions are used to fetch/create/update profile from the dynamodb database.

### CODE REVIEW

File: template.yaml [hunk 1]
```diff
@@ -81,6 +81,28 @@ Resources:
...
+            Method: POST
+      Environment:
+        Variables:
+          DYNAMODB_TABLE: !Ref ProfileTable
+      Policies:
+        - DynamoDBWritePolicy:
+            TableName: !Ref ProfileTable
+  
+  ProfileTable:
+    Type: AWS::Serverless::SimpleTable
```
[the remark this answers is not in the document]
Comment (midepeter) [reply template.yaml:104]: Okay Noted I will correct that

**Comment by midepeter** [comment]
Okay. Thanks for this

**Comment by midepeter** [comment]
Alright

File: template.yaml [hunk 2]
```diff
@@ -146,6 +145,64 @@ Resources:
...
+              Authorizer: "NONE"
+      Environment:
+        Variables:
+          PROFILE_TABLE: !Ref ProfileTable
+      Policies:
+        - DynamoDBCrudPolicy:
+            TableName:
+              !Ref ProfileTable
+
+  CreateProfileHandlerFunction:
```
[the remark this answers is not in the document]
Comment (gregoflash05) [reply template.yaml:169]: I actually cannot do this, I tried for hours and it didn't work

File: profile-crud/main.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	jsonData, _ := json.Marshal(msg)
+
+	return events.APIGatewayProxyResponse{
+		StatusCode: http.StatusOK,
+		Body:       string(jsonData),
+	}, err
+}
+
+func getProfileByID(req events.APIGatewayV2HTTPRequest) (events.APIGatewayProxyResponse, error) {
```
Comment (rubenhorn) [inline profile-crud/main.go:58]: This seems like a lot of duplicate code.
Can this be merged with the other get-function?

File: profileservice/dynamodb.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("!! Done updating profile !!") // TODO remove
+
+	return nil
+}
+
+func UpdateProfileTableItem(i string, v string, userID string, svc *dynamodb.DynamoDB, ch chan error, wg *sync.WaitGroup) {
```
Comment (rubenhorn) [inline profileservice/dynamodb.go:99]: We should update all attributes at the same time: https://stackoverflow.com/a/51072244
(Then we also don't need the goroutines and synchronization mechanism)

Comment (gregoflash05) [reply profileservice/dynamodb.go:99]: This was the only safe way to wrote the update function, and also make it fast.

Comment (gregoflash05) [reply profileservice/dynamodb.go:99]: The reference is written in JavaScript.

Comment (gregoflash05) [reply profileservice/dynamodb.go:99]: I tried updating at once but this became a problem "UpdateExpression: aws.String("set " + i + " = :r"),". I needed a way to generate the update statement (which is more like the query language) and make it generic. In the end, I still needed to loop. This solution made it less complex.

**Comment by zen37** [comment]
**models.go**

- have same validation length for both update and create

- perhaps define constants with the validation requirements (parameters would be better but they might be an overkill as they should not change often)


**controllers.go**

- session

	sess := session.Must(session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
	}))
	vs
	sess = session.Must(session.NewSession())

- skipFields
	var skipFields = []string{"userid", "displayname", "phone"}
	Displayname should be allowed to be updated
	use parameter store for skipFields, again it could be an overkill

- variable _nw_
why nw :) ? if the name is very dear to Chigozie it can stay like that, and future generations will try to solve the puzzle, perhaps it has a hidden meaning that will unlock a treasure map :)

- func contains(v string, a []string) bool {
this might be needed in other parts of the application, should we have a utilities.go file?


**main.go**	
- Message: "Error Retreiving Profile",
just spelling

-func handler(
suggested switch/case instead of else if, but Chigozie does not like it, not that he hates switch with every fibre of his body, he just does not like switch and he very much prefers else if :).

**Comment by zen37** [comment]
**aggregator/main.go**
```
	for k, v := range aggregations {
		// TODO implement
		fmt.Printf("Change participation count of %s by %d\n", k, v)
```
anything else to implement here?

**api/main.go**
`func doesParticipationExist(`
wouldn't it be better to be called `participationExists(`

`	json, _ := json.Marshal(participation.ParticipationEvent{`
return error, very unlikely an error will happen here but just to be consistent

`var err error`
any advantage declaring the variable in advance, one line fewer if we don't :)

**Comment by gregoflash05** [comment]
@zen37 I have made changes based on the comments

**Review by zen37 — APPROVED** [review]

**Pull request merged**